### PR TITLE
backend: k8cache: Fix panic in DeleteKeys via ignored GenerateKey error

### LINKED ISSUE #5214: Fatal panic in `DeleteKeys` when `GenerateKey` returns error for short URL paths
## Bug

When caching is enabled (`--cache-enabled=true`), a non-GET request to a short API path (e.g., `/clusters/<name>`) causes an unrecoverable `runtime error: index out of range [2] with length 1` panic in `DeleteKeys`, crashing the entire backend process.

## Root Cause

`HandleNonGETCacheInvalidation` discards the error from `GenerateKey` with `_`, then passes the empty key `""` to `DeleteKeys`. Inside `DeleteKeys`, `strings.Split("", "+")` returns `[""]` (length 1), and `keyPart[2] = ""` panics with index out of range.

```go
// cacheInvalidation.go:67-68 (before fix)
key, _ := GenerateKey(r.URL, contextKey)  // error dropped, key = ""
DeleteKeys(key, k8scache)                  // panics inside
```

## Impact

- Complete backend process crash (no `recover()` in path)
- All WebSocket connections and HTTP handlers terminated
- Requires manual server restart

## Note

This is independent from PR #5161 which fixes OOB panics inside `GetAPIGroup`. The `len(parts) < 4` early-return guard that triggers this bug is preserved in #5161.

### PR DESCRIPTION
**Pull request by KumarADITHYA123**
Fixes #5214 

## Summary

`HandleNonGETCacheInvalidation` discards the error from `GenerateKey` with `_`, passing an empty key to `DeleteKeys`, which panics on `keyPart[2]` (index out of range). This crashes the entire backend when caching is enabled and a non-GET request hits a short API path.

## Changes

- **`cacheInvalidation.go`**: Handle `GenerateKey` error instead of discarding it; add `len(keyPart) < 3` guard in `DeleteKeys`
- **`cacheInvalidation_test.go`**: Add 2 regression tests for empty/malformed keys

## Steps to Test

```bash
cd backend && go test ./pkg/k8cache/... -v -count=1
```

All 35 tests pass (including 2 new regression tests).

## Notes for the Reviewer

- Independent from PR #5161 (different crash site, different fix location)
- Surgical fix: +9/-2 lines of logic, zero refactoring
- Request continues to forward to K8s API even when cache invalidation is skipped

### CODE REVIEW

**Comment by KumarADITHYA123** [comment]
"Note: The failure in the s390x CI run is a pre-existing path separator (\ vs /) issue in plugins_test.go and is unrelated to the k8cache changes in this PR."

**Review by skoeva — CHANGES_REQUESTED** [review]
the s390x test failure is isolated to this PR, have you verified that it's unrelated?

**Comment by KumarADITHYA123** [comment]
Yes, I've completely verified that this failure is unrelated to this PR.

1. **Isolation:** The only files modified in this PR are `backend/pkg/k8cache/cacheInvalidation.go` and `backend/pkg/k8cache/cacheInvalidation_test.go`. There are zero changes or dependencies related to `backend/pkg/plugins/`.
2. **Pre-existing Flake:** The `TestWatch` test in `plugins_test.go` (specifically at line 87) is failing due to a cross-platform path separator mismatch (`\` vs `/`). The test expects `.../e381...:CREATE` but the filesystem watcher emits `...\e381...:CREATE`. 
3. **Reproducible on `main`:** I checked `upstream/main`, and this exact test failure happens there as well on certain environments like the `s390x` emulation.
4. **k8cache tests passed:** All tests for `k8cache` successfully passed in this exact CI run (you can see `github.com/kubernetes-sigs/headlamp/backend/pkg/k8cache` passed in the logs right above the `plugins` failure).

So the `s390x` failure is a pre-existing path logic issue in `plugins_test.go` and is completely isolated from the caching changes in this PR.

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for these changes.


Can you please address the open review comments?

File: backend/pkg/k8cache/cacheInvalidation.go [hunk 1]
```diff
@@ -64,8 +69,12 @@ func HandleNonGETCacheInvalidation(k8scache cache.Cache[string], w http.Response
...
 	}
 
-	key, _ := GenerateKey(r.URL, contextKey)
-	DeleteKeys(key, k8scache)
+	key, err := GenerateKey(r.URL, contextKey)
+	if err != nil {
+		logger.Log(logger.LevelWarn, nil, err, "could not generate cache key; skipping invalidation")
+	} else {
+		DeleteKeys(key, k8scache)
+	}
```
[the remark this answers is not in the document]
Comment (KumarADITHYA123) [reply backend/pkg/k8cache/cacheInvalidation.go]: Good catch. I've updated HandleNonGETCacheInvalidation to return early with ErrHandled when GenerateKey fails, preventing the flow from continuing and incorrectly storing data under an empty cache key.

File: backend/pkg/k8cache/cacheInvalidation.go [hunk 2]
```diff
@@ -47,7 +47,12 @@ import (
 // also empty namespace.
 func DeleteKeys(key string, k8scache cache.Cache[string]) {
 	_ = k8scache.Delete(context.Background(), key)
+
 	keyPart := strings.Split(key, "+")
+	if len(keyPart) < 3 {
+		return // malformed key; nothing to namespace-strip
+	}
+
 	keyPart[2] = ""
```
[the remark this answers is not in the document]
Comment (KumarADITHYA123) [reply backend/pkg/k8cache/cacheInvalidation.go:56]: Fixed. I updated the boundary guard to len(keyPart) < 4 to strictly enforce the 4-part canonical key format (apiGroup+kind+namespace+context).

File: backend/pkg/k8cache/cacheInvalidation_test.go [hunk 3]
```diff
@@ -81,6 +81,34 @@ func TestDeleteKeys(t *testing.T) {
...
+					"+pods+default+test-context": "value-1",
+				},
+			},
+			key: "partial+key",
+			aftermockCache: &MockCache{ //nolint:exhaustruct
+				store: map[string]string{
+					"+pods+default+test-context": "value-1",
+				},
+			},
+		},
```
[the remark this answers is not in the document]
Comment (KumarADITHYA123) [reply backend/pkg/k8cache/cacheInvalidation_test.go:111]: Added! I included a new regression test specifically for an exactly 3-part key (group+kind+ns) to ensure it gets correctly flagged as malformed and safely ignored.

**Comment by KumarADITHYA123** [comment]
@illume I've addressed all 3 open Copilot review comments regarding the cache invalidation edge cases. Let me know if everything looks good!

**Review by illume — APPROVED** [review]
Thanks! 🎉
